lib/spack/spack/preferred_packages.py

```diff
@@ -154,7 +154,10 @@ def spec_has_preferred_provider(self, pkgname, provider_str):
 
     def spec_preferred_variants(self, pkgname):
         """Return a VariantMap of preferred variants and their values"""
-        variants = self.preferred.get(pkgname, {}).get('variants', '')
+        for pkg in (pkgname, 'all'):
+            variants = self.preferred.get(pkg, {}).get('variants', '')
```

Comment:
This is working correctly as far as I can tell. I have added a dummy variant to `m4` and built it with and without an appropriate `packages.yaml`. The hash of the only dependency does not change and both `m4` variants use the same dependency.